File: shipbob_to_gorgias/csharp/sb2gorgias/ApiClient.cs
using System.Diagnostics;
using System.Net;
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json.Nodes;

using Microsoft.Extensions.Logging;

namespace Sb2Gorgias;

/// <summary>
/// A small JSON-over-HTTP client shared by both API clients: standing headers plus a
/// bounded retry. Pandium does not retry a failed run, so a transient 429 or gateway error
/// has to be absorbed here or the whole run is lost.
/// </summary>
public sealed class ApiClient : IDisposable
{
    /// <summary>Rate limiting, plus the gateway errors both APIs return under load.</summary>
    private static readonly HashSet<HttpStatusCode> RetryStatuses =
    [
        HttpStatusCode.TooManyRequests,
        HttpStatusCode.BadGateway,
        HttpStatusCode.ServiceUnavailable,
        HttpStatusCode.GatewayTimeout,
    ];

    /// <summary>Total attempts, the first included.</summary>
    private const int MaxAttempts = 6;

    /// <summary>
    /// The longest a <c>Retry-After</c> is honoured for. A client that sleeps past Pandium's
    /// run limit never reaches the stdout write that ends the run successfully.
    /// </summary>
    private static readonly TimeSpan MaxRetryAfter = TimeSpan.FromSeconds(60);

    private readonly HttpClient _http;
    private readonly TimeSpan _backoff;
    private readonly ILogger _logger;

    /// <param name="baseUrl">The API root. Every path below is relative to it.</param>
    /// <param name="authorization">The standing <c>Authorization</c> header.</param>
    /// <param name="backoff">How long to wait before the first retry; doubled before each one after that.</param>
    /// <param name="logger">Where retries are announced.</param>
    public ApiClient(string baseUrl, AuthenticationHeaderValue authorization, TimeSpan backoff, ILogger logger)
    {
        _http = new HttpClient
        {
            // The trailing slash makes relative paths append to the API root rather than
            // replace its last segment.
            BaseAddress = new Uri(baseUrl.TrimEnd('/') + "/"),
        };
        _http.DefaultRequestHeaders.Authorization = authorization;
        _http.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
        _backoff = backoff;
        _logger = logger;
    }

    public Task<JsonNode?> GetAsync(string path, (string Key, string Value)[] query, CancellationToken token) =>
        SendAsync(() => new HttpRequestMessage(HttpMethod.Get, path + QueryString(query)), token);

    public Task<JsonNode?> PostAsync(string path, JsonNode body, CancellationToken token) =>
        SendAsync(() => new HttpRequestMessage(HttpMethod.Post, path) { Content = JsonContent(body) }, token);

    public Task<JsonNode?> PutAsync(string path, JsonNode body, CancellationToken token) =>
        SendAsync(() => new HttpRequestMessage(HttpMethod.Put, path) { Content = JsonContent(body) }, token);

    public void Dispose() => _http.Dispose();

    private static StringContent JsonContent(JsonNode body) =>
        new(body.ToJsonString(), Encoding.UTF8, "application/json");

    private static string QueryString((string Key, string Value)[] query) =>
        query.Length == 0
            ? ""
            : "?" + string.Join('&', query.Select(item => $"{item.Key}={Uri.EscapeDataString(item.Value)}"));

    /// <summary>
    /// Send until the response is not a retryable status, then parse the body as JSON.
    ///
    /// A non-2xx that survives the retries throws, carrying the status and the body. A
    /// failure is never mapped onto an empty result: the cron flow commits its cursor on an
    /// empty page, so the two have to stay distinct.
    /// </summary>
    /// <param name="newRequest">
    /// Builds a fresh request for each attempt; an <see cref="HttpRequestMessage"/> cannot
    /// be sent twice.
    /// </param>
    private async Task<JsonNode?> SendAsync(Func<HttpRequestMessage> newRequest, CancellationToken token)
    {
        var wait = _backoff;
        for (var attempt = 1; attempt <= MaxAttempts; attempt++)
        {
            using var request = newRequest();
            using var response = await _http.SendAsync(request, token);

            if (RetryStatuses.Contains(response.StatusCode) && attempt < MaxAttempts)
            {
                var pause = RetryAfter(response) ?? wait;
                _logger.LogWarning(
                    "HTTP {Status} from {Uri}; retrying in {Pause} (attempt {Attempt})",
                    (int)response.StatusCode, request.RequestUri, pause, attempt);
                await Task.Delay(pause, token);
                wait *= 2;
                continue;
            }

            var body = await response.Content.ReadAsStringAsync(token);
            if (!response.IsSuccessStatusCode)
            {
                throw new HttpRequestException(
                    $"HTTP {(int)response.StatusCode} from {request.RequestUri}: {body}");
            }

            // A 204, and some 200s, come back empty; null lets callers index into the
            // result either way.
            return string.IsNullOrWhiteSpace(body) ? null : JsonNode.Parse(body);
        }

        throw new UnreachableException("the retry loop returns or retries on every attempt");
    }

    /// <summary>
    /// How long the response asked the client to wait, or null to fall back to the doubling
    /// backoff. Clamped so a date already passed reads as "now" and a long wait cannot
    /// outlive the run.
    /// </summary>
    private static TimeSpan? RetryAfter(HttpResponseMessage response)
    {
        if (response.Headers.RetryAfter is not { } header)
        {
            return null;
        }

        if ((header.Delta ?? (header.Date - DateTimeOffset.UtcNow)) is not { } wait)
        {
            return null;
        }

        return TimeSpan.FromTicks(Math.Clamp(wait.Ticks, 0, MaxRetryAfter.Ticks));
    }
}

File: shipbob_to_gorgias/java/src/main/java/sb2gorgias/ApiClient.java
package sb2gorgias;

import java.io.IOException;
import java.io.UncheckedIOException;
import java.net.URI;
import java.net.URLEncoder;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.Map;
import java.util.Optional;
import java.util.Set;

import org.json.JSONException;
import org.json.JSONObject;
import org.json.JSONTokener;

/** A small HTTP client with exponential backoff retry, scoped to one API's base URL and auth
 * header. requester is swappable so tests can stub the network without a full HTTP mock. */
final class ApiClient {
    private static final Set<Integer> RETRY_STATUSES = Set.of(429, 502, 503, 504);
    private static final int MAX_ATTEMPTS = 6;

    // A server-requested wait longer than this is not worth honoring - it would tie up the run
    // far past what our own backoff schedule ever would anyway.
    private static final Duration MAX_RETRY_AFTER = Duration.ofSeconds(60);

    private static final HttpClient HTTP = HttpClient.newBuilder()
            .connectTimeout(Duration.ofSeconds(10))
            .build();

    /** One raw HTTP call, decoupled from the JDK client so tests can fake it directly. */
    interface RawRequester {
        RawResponse send(String method, String url, String body) throws IOException;
    }

    /** retryAfter is the server's Retry-After value, in seconds form (the only form these
     * APIs are expected to send) - present only on a response that actually carried one. */
    record RawResponse(int status, String body, Optional<Duration> retryAfter) {
        RawResponse(int status, String body) {
            this(status, body, Optional.empty());
        }
    }

    private final String baseUrl;
    private final String authorization;
    private final Duration backoff;
    private final Set<String> retryMethods;
    RawRequester requester;

    ApiClient(String baseUrl, String authorization, Duration backoff, Set<String> retryMethods) {
        this.baseUrl = baseUrl;
        this.authorization = authorization;
        this.backoff = backoff;
        this.retryMethods = retryMethods;
        this.requester = this::jdkSend;
    }

    String authorization() {
        return authorization;
    }

    Object get(String path, Map<String, String> query) {
        StringBuilder url = new StringBuilder(baseUrl).append(path);
        if (query != null && !query.isEmpty()) {
            url.append('?');
            boolean first = true;
            for (Map.Entry<String, String> entry : query.entrySet()) {
                if (!first) {
                    url.append('&');
                }
                first = false;
                url.append(encode(entry.getKey())).append('=').append(encode(entry.getValue()));
            }
        }
        return send("GET", url.toString(), null);
    }

    Object post(String path, JSONObject body) {
        return send("POST", baseUrl + path, body == null ? null : body.toString());
    }

    Object put(String path, JSONObject body) {
        return send("PUT", baseUrl + path, body == null ? null : body.toString());
    }

    /** Retries a retryable failure with exponential backoff, or the server's own Retry-After
     * when it sends one. Stops immediately if the backoff sleep is interrupted. An empty body
     * returns null, not an error, so callers can tell "nothing here" from a real failure. */
    private Object send(String method, String url, String body) {
        RuntimeException lastError = null;
        Duration retryAfter = null;
        for (int attempt = 1; attempt <= MAX_ATTEMPTS; attempt++) {
            if (attempt > 1) {
                Duration wait = retryAfter != null ? retryAfter : backoff.multipliedBy(1L << (attempt - 2));
                retryAfter = null;
                if (!sleep(wait)) {
                    throw lastError;
                }
            }

            RawResponse response;
            try {
                response = requester.send(method, url, body);
            } catch (IOException e) {
                lastError = new UncheckedIOException(e.getMessage(), e);
                if (retryMethods.contains(method)) {
                    continue;
                }
                throw lastError;
            }

            if (response.status() >= 400) {
                lastError = new RuntimeException("HTTP " + response.status() + ": " + response.body());
                if (RETRY_STATUSES.contains(response.status()) && retryMethods.contains(method)) {
                    retryAfter = response.retryAfter().orElse(null);
                    continue;
                }
                throw lastError;
            }

            if (response.body() == null || response.body().isBlank()) {
                return null;
            }
            try {
                return new JSONTokener(response.body()).nextValue();
            } catch (JSONException e) {
                throw new RuntimeException("could not decode response body: " + e.getMessage(), e);
            }
        }
        throw lastError;
    }

    private RawResponse jdkSend(String method, String url, String body) throws IOException {
        HttpRequest.BodyPublisher publisher = body == null
                ? HttpRequest.BodyPublishers.noBody()
                : HttpRequest.BodyPublishers.ofString(body, StandardCharsets.UTF_8);
        HttpRequest request = HttpRequest.newBuilder(URI.create(url))
                .timeout(Duration.ofSeconds(30))
                .header("accept", "application/json")
                .header("content-type", "application/json")
                .header("Authorization", authorization)
                .method(method, publisher)
                .build();

        try {
            HttpResponse<String> res = HTTP.send(request, HttpResponse.BodyHandlers.ofString());
            return new RawResponse(res.statusCode(), res.body(), retryAfterOf(res));
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            throw new IOException(e);
        }
    }

    /** Only the seconds form of Retry-After is handled; anything else (an HTTP-date) is
     * treated as absent rather than failing the request over it. */
    private static Optional<Duration> retryAfterOf(HttpResponse<?> res) {
        return res.headers().firstValue("Retry-After").flatMap(value -> {
            try {
                long seconds = Long.parseLong(value.trim());
                return seconds > 0
                        ? Optional.of(Duration.ofSeconds(Math.min(seconds, MAX_RETRY_AFTER.toSeconds())))
                        : Optional.<Duration>empty();
            } catch (NumberFormatException e) {
                return Optional.empty();
            }
        });
    }

    private static String encode(String s) {
        return URLEncoder.encode(s, StandardCharsets.UTF_8);
    }

    /** Returns false if interrupted mid-sleep, restoring the interrupt flag either way. */
    private static boolean sleep(Duration d) {
        try {
            Thread.sleep(d.toMillis());
            return true;
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            return false;
        }
    }
}
